lib/spack/spack/hooks/check_dynamic_linking_elf.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import struct
+
+import llnl.util.tty as tty
+from llnl.util.lang import elide_list
+
+import spack.error
+import spack.object_file.elf as elf
+
+skip_list = set([
+    # kernel
+    'linux-vdso',
+    # glibc
+    'ld-linux-x86-64',
+    'libc',
+    'libdl',
+    'libgcc_s',
+    'libm',
+    'libmemusage',
+    'libmvec',
+    'libnsl',
+    'libnss_compat',
+    'libnss_db',
+    'libnss_dns',
+    'libnss_files',
+    'libnss_hesiod',
+    'libpcprofile',
+    'libpthread',
+    'libresolv',
+    'librt',
+    'libSegFault',
+    'libthread_db',
+    'libutil',
+    # gcc
+    'libasan',
+    'libatomic',
+    'libcc1',
+    'libgcc_s',
+    'libgfortran',
+    'libgomp',
+    'libitm',
+    'liblsan',
+    'libquadmath',
+    'libssp',
+    'libstdc++',
+    'libtsan',
+    'libubsan',
+])
+
+
+def is_compatible(elf_a, elf_b):
+    return (elf_a.is_little_endian == elf_b.is_little_endian and
+            elf_a.is_64_bit == elf_b.is_64_bit and
+            elf_a.isa == elf_b.isa)
+
+
+def candidate_matches(current_elf, candidate_path):
+    try:
+        with open(candidate_path, 'rb') as g:
+            return is_compatible(current_elf, elf.parse_elf_file(g))
+    except (IOError, DeprecationWarning, struct.error, elf.ElfParsingError):
+        return False
+
+
+def read_elf(path):
+    try:
+        with open(path, 'rb') as f:
+            current_elf = elf.parse_elf_file(f)
+            if not current_elf.dynamic or len(current_elf.needed_offsets) == 0:
+                return None
+            needed = [
+                elf.read_c_string_from_file(f, off).decode("utf-8") for
+                off in current_elf.needed_offsets
+            ]
+            return current_elf, needed
+    except (IOError, DeprecationWarning, struct.error, elf.ElfParsingError):
+        return None
+
+
+def locate_needed(search_dirs, current_elf, needed, num_not_found):
+    """
+    Find libraries by <search dir>/<name>. At the end the needed
+    list is partitioned as [not found | found] and the number of
+    libraries that are not found is returned.
+    """
+    for dir in search_dirs:
+        i = 0
+        while i < num_not_found:
+            # Try to open by path
+            candidate = os.path.join(dir, needed[i])
+
+            if candidate_matches(current_elf, candidate):
+                num_not_found -= 1
+                needed[i], needed[num_not_found] = needed[num_not_found], needed[i]
+            else:
+                i += 1
+
+    return num_not_found
+
+
+def partition_needed(needed):
+    """
+    Partition the needed list in [soname | path/skipped]
+    """
+    num_sonames = len(needed)
+    i = 0
+    while i < num_sonames:
+        if '/' in needed[i] or needed[i].split(".so", 1)[0] in skip_list:
```

Comment:
`'/' in needed[i]` this would be clearer if you created a function called `is_file_path(...)` (also on Windows paths have `\`).

Reply:
Right, but this is not portable and rather specific to DT_NEEDED (https://github.com/bminor/glibc/blob/d3e4f5a1014db09ff1c62c6506f92cba469e193d/elf/dl-load.c#L2135). I was thinking to keep `needed` strings simply byte strings, since that's what paths are on linux anyways, and then check for presence of `b'/'`. But it's hard to not use utf-8 strings it seems...